src/TinyUPnP.cpp

```diff
@@ -484,51 +438,345 @@ boolean TinyUPnP::waitForUnicastResponseToMSearch(gatewayInfo *deviceInfo, IPAdd
+				tagContent =  getTagContent(line, "NewInternalPort");
+				if(tagContent != ""){
+					rule_ptr->internalPort = tagContent.toInt();
+					tagContent = "";
+					paramFlag |= 0x08;
+					UPNP_DEBUGln("Found NewInternalPort");
+				}
+			}
+			if( !(paramFlag & 0x10) && (paramFlag & 0x01) ){	//Check for NewLeaseDuration
+				tagContent =  getTagContent(line, "NewLeaseDuration");
+				if(tagContent != ""){
+					rule_ptr->leaseDuration = tagContent.toInt();
+					tagContent = "";
+					paramFlag |= 0x10;
+					UPNP_DEBUGln("Found NewLeaseDuration");
+				}
+			}
+			if(paramFlag == 0x1F){
+				UPNP_DEBUGln("Response OK");
+				response = UPNP_OK;
+			}
+			/*if (line.indexOf((SOAPActions[GetSpecificPortMappingEntry]+String("Response"))) >= 0){
+				rule_ptr->devFriendlyName = getTagContent(line, "NewPortMappingDescription");
+				String newInternalClient = getTagContent(line, "NewInternalClient");
+				rule_ptr->internalAddr.fromString(newInternalClient);
+				rule_ptr->internalPort = getTagContent(line, "NewInternalPort").toInt();
+				rule_ptr->leaseDuration = getTagContent(line, "NewLeaseDuration").toInt();
+				response = UPNP_OK;
+			}*/
+		}else if (SOAPAction == DeletePortMapping && response != UPNP_OK){
+			if (line.indexOf((SOAPActions[DeletePortMapping]+String("Response"))) >= 0){
+				response = UPNP_OK;
+			}
+		}else if(SOAPAction == GetExternalIPAddress && response != UPNP_OK){
+			if ( !(paramFlag & 0x01) && line.indexOf((SOAPActions[GetExternalIPAddress]+String("Response"))) >= 0){
+				paramFlag |= 0x01;	// Response OK
+				UPNP_DEBUGln("Response recived");
+			}
+			if( !(paramFlag & 0x02) && (paramFlag & 0x01) ){	//Check for NewExternalIPAddress
+				tagContent = getTagContent(line, "NewExternalIPAddress");
+				if(tagContent != ""){
+					rule_ptr->internalAddr.fromString(tagContent);
+					paramFlag |= 0x02;
+					UPNP_DEBUGln("Found NewExternalIPAddress");
+				}
+			}
+			if(paramFlag == 0x03){
+				UPNP_DEBUGln("Response OK");
+				response = UPNP_OK;
+			}
+			/*if (line.indexOf((SOAPActions[GetExternalIPAddress]+String("Response"))) >= 0){
+				String NewExternalIPAddress = getTagContent(line, "NewExternalIPAddress");
+				rule_ptr->internalAddr.fromString(NewExternalIPAddress);
+				response = UPNP_OK;
+			}*/
+		}
+		if (response == UPNP_OK || response == INVALID_INDEX || response == 714 || line.indexOf(F("</s:Envelope>")) >= 0){
+			UPNP_DEBUG("\nEnd of Response or Flush due to Response:");
+			UPNP_DEBUGln(response);
+			while (_wifiClient.available()) {
+				UPNP_DEBUG((char)_wifiClient.read());
+			}
+			break;
+		}
+	}
+	return response;
+}
+
+IPAddress TinyUPnP::getExternalIP(){
+	IPAddress myIP; 
+	_upnpRule *result_ptr = new _upnpRule();
+	upnpResult result = postSOAPAction(&_gwInfo, result_ptr, GetExternalIPAddress);
+	if (result != UPNP_OK){
+		UPNP_DEBUG("Error:");
+		UPNP_DEBUGln(String(result));
+		delete result_ptr;
+		return false;
```

Comment:
:/ return type

Reply:
oh, Totally missed it .. thanks, I'll change it ASAP ..
also don't mind the Commented parts of the code .. I will remove them for the final Pull but after complete testing ..